### DISCUSSION TITLE
My legacy (2.12.17) Scala code no longer compiles

### DISCUSSION THREAD

**Discussion by ghost** [scalameta/metals#5468] [General]
Hello all,

I just got a Metals update (v1.24.0) in VSCode (Fedora Linux 38). My Scala code is quite old and my build.sbt contains 

scalaVersion := "2.12.17"

which used to work fine. Now Metals/SBT seems to use the wrong compiler and library (scala-library-2.13.10.jar is displayed). My code no longer compiles and gives my 600+ errors.

Please help!

**Comment by tgodzik** [comment]
Nothing should have really changed as far as compilation is concerned, we haven't removed any support. Can you run the doctor command? Are you able to import the build again or clean compile? Anything in the Metals output? Any details would be helpful.

**Reply by ghost** [reply]
Strangely the output says that Metals is version 1.0.0. The terminal output show Bloop is at version v1.5.8. The output starts with:

2023.07.20 17:05:55 INFO  Attempting to connect to the build server...
2023.07.20 17:05:55 WARN  Build server is not auto-connectable.
2023.07.20 17:05:55 WARN  no build tool detected in workspace '/home/silvio/java_root/projects/scripture/build/scripts'. The most common cause for this problem is that the editor was opened in the wrong working directory, for example if you use sbt then the workspace directory should contain build.sbt. 
2023.07.20 17:05:55 WARN  scala-cli >= 0.1.3 not found in PATH, fetching and starting a JVM-based Scala CLI
2023.07.20 17:05:55 INFO  Bloop uses /usr/lib/jvm/java-20-openjdk-20.0.1.0.9-8.rolling.fc38.x86_64 defined at /home/silvio/.bloop/bloop.json
2023.07.20 17:05:55 INFO  skipping build import with status 'Installed'
2023.07.20 17:05:55 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at /home/silvio/java_root/projects/scripture/.metals/bsp.trace.json or /home/silvio/.cache/metals/bsp.trace.json
2023.07.20 17:05:55 INFO  Attempting to connect to the build server...
2023.07.20 17:05:55 INFO  Bloop uses /usr/lib/jvm/java-20-openjdk-20.0.1.0.9-8.rolling.fc38.x86_64 defined at /home/silvio/.bloop/bloop.json
2023.07.20 17:05:55 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at /home/silvio/java_root/projects/scripture/project/.metals/bsp.trace.json or /home/silvio/.cache/metals/bsp.trace.json
2023.07.20 17:05:55 INFO  time: Connected to build server in 0.38s

**Reply by ghost** [reply]
When I try import build (which used to work normally) I get a prompt to pick a folder that is inside my workspace. This never happened before.
When I pcik one of the project folders it says: No workspace folder chosen to execute import build.

**Reply by ghost** [reply]
When starting SBT from the command line in the project folder acompilation suceeds. But in VSCode metals still picks the Scala 2.13.1 library and gives 6K+ errors. The "Procedure syntax not supported" error is everywhere.

**Reply by tgodzik** [reply]
Are you using multiple folders in the workspace? We only now introduced the support for it, but looks like we haven't checked it on a setup remotely similar to yours. In the folder that contains the sbt project, is it opened within the folder that contains build.sbt? It seems we are unable to detect that.

**Reply by ghost** [reply]
The project folder contains the workspace-file the sbt file. The sources are inside the same folder. So:

project/build.sbt
project/project.code-workspace
project/src/main/scala/nl/mycompany/folder1
project/src/main/scala/nl/mycompany/folder2
project/src/main/scala/nl/mycompany/folder3

I have to add folder1, folder2 and folder3 to the workspace becuase the SVN plugin will not pick op the three repos otherwise.

**Reply by ghost** [reply]
Is there any way to revert to the previous version? I am close to a deadline and can not afford to spend time on this.

**Reply by ghost** [reply]
Yes, this is where it breaks. If I make a completely new workspace and do NOT add the subfolders to the workspace things work fine. As soon as I add the subfolders Metals starts compiling with the wrong Scala version and thing are broken.
This used to work fine with the previous version. The added subfolder where displayed and compiled with the same setting as the containing projects build.sbt specifies.

Unfortunately I need the subfolders or my SVN plugin will not pickup on the repos.

**Reply by ghost** [reply]
Final note: the SVN plugin now supports picking up on the repos without adding them as folders in the workspace. Not as nice as it was but at least I have a working environment now.

Thanks for the help.

**Reply by tgodzik** [reply]
In the previous versions we would pick up only the first folder and I guess that would be the one with `build.sbt`. Since the other folders were inside the first one this must have been handled correctly by sheer luck and that use case not take into account on purpose at all.

We assume that workspace folders would be separate from each other and I am not entirely sure how to handle this situation :thinking: 
Maybe we should add a heuristic that melds the workspaces together? But this might be hard to achieve, but you can report it as a feature request.

For the future, you can always revert the extension version in the extension tab if you are having difficulties with the newer version.
